File: leetcode/trie/336. Palindrome Pairs/py/main.py
# Detailed solution: https://leetcode.com/problems/palindrome-pairs/solution/

class Solution:
  def palindromePairs(self, words):
    trie = Trie(words)
    res = []

    for i in range(len(words)):
      pairs = trie.search(words[i], i)
      res = res + pairs
    
    return res

class Node :
  def __init__(self):
    self.children = {}
    self.ending_word = -1
    self.palindrome_suffixes = []

class Trie:
  def __init__(self, words):
    self.root = Node()

    for i in range(len(words)):
      self.insert(words[i], i)

    # Put all the words into the Trie in reverse
  def insert(self, word, index):
    current = self.root

    for i in reversed(range(len(word))):
      if word[i] not in current.children:
        current.children[word[i]] = Node()

      # Case 2 Reversed word + palindrome  
      # Rest of the word is a palindrome 
      if self.is_palidrome(word, 0, i):
        current.palindrome_suffixes.append(index)

      current = current.children[word[i]]
      
    current.ending_word = index
    current.palindrome_suffixes.append(index)

  # Case 1 is where a palindrome pair is formed by 2 words that are the reverse of each other. 
  # Remember that for the output, we need to give the indexes of each pair. 
  # We'll add an index field onto each end of word node.
  def search(self, word, index):
    current = self.root
    pairs = []

    for i in range(len(word)):
      # Case 3 when first word greater than second word in the trie 
      # And prefix first word == second word and rest suffix of current word is a palindrome
      
      # Example: word1: "CATSOLOS" word2: "TAC" where CAT == TAC and SOLOS is palindrome
      
      if current.ending_word >= 0 and current.ending_word != index and self.is_palidrome(word, i, len(word) - 1):
        pairs.append([index, current.ending_word])
      
      if word[i] not in current.children:
        return pairs
      
      current = current.children[word[i]]

    # Case 2 current word == reversed word in the trie and rest is a palindrome 
    # Example: word1: "CATSOLOS" word2: "TAC" where CAT == TAC and SOLOS is palindrome 
    for i in current.palindrome_suffixes:
      if i != index:
        pairs.append([index, i])

    return pairs

  def is_palidrome(self, word, lo, hi):
    while lo < hi:
      if word[lo] != word[hi]:
        return False
      lo += 1
      hi -= 1
    return True

File: leetcode/trie/336. Palindrome Pairs/js/main.js
/**
 * Solution using Trie
 * Time complexity: O(n * k^2) where k is Where n is the number of words in the list
 * and k is the maximum length that is checked for palindrome.
 *
 * An intuitive brute force solution time complexity is O(n^2 * k).
 */

// Time O(N * k^2)
// Space O(N)
const palindromePairs = words => {
  let results = [];
  let trie = new Trie(words);

  words.forEach((word, index) => {
    let pairs = trie.search(word, index);
    results = results.concat(pairs);
  });

  return results;
};

class TrieNode {
  constructor() {
    this.children = {};
    this.index = -1;
    this.palindromeIndices = [];
  }
}

class Trie {
  constructor(words) {
    this.root = new TrieNode();

    words.forEach((word, index) => {
      this.insert(word, index);
    });
  }

  insert(word, index) {
    let current = this.root;

    for (let i = word.length - 1; i >= 0; i--) {
      const c = word[i];

      if (!current.children[c]) {
        current.children[c] = new TrieNode();
      }

      if (isPalindrome(word, 0, i)) {
        current.palindromeIndices.push(index);
      }

      current = current.children[c];
    }

    current.index = index;
    current.palindromeIndices.push(index);
  }

  search(word, index) {
    let pairs = [];

    let current = this.root;

    for (let i = 0; i < word.length; i++) {
      // Case 1. слово длинное
      // 0:    abc
      //       |
      //       index = 0, palindromeIndices: [0]
      //       |
      //       current
      //       |
      // 1: cbaaaaaaa (word)
      //       i
      //
      // [1, 0] is a pair

      // На проверку полидрома уйдет слово aaaaaa оно полидром
      if (current.index >= 0 && current.index !== index && isPalindrome(word, i, word.length - 1)) {
        pairs.push([index, current.index]);
      }

      if (!current.children[word[i]]) {
        return pairs; // Mismatch
      }

      current = current.children[word[i]];
    }

    // Case 2. длинное слово
    // 0: aaaaaaabc
    //          |
    //          palindromeIndices = [0]
    //          |
    //          current
    //          |
    // 1:    cba
    //          i
    current.palindromeIndices.forEach(i => {
      if (i !== index) {
        pairs.push([index, i]);
      }
    });

    return pairs;
  }
}

const isPalindrome = (str, i, j) => {
  while (i < j) {
    if (str[i++] !== str[j--]) {
      return false;
    }
  }
  return true;
};

// Time O(N^2 * K)
const palindromePairs_II = words => {
  let res = [];

  for (let i = 0; i < words.length; i++) {
    for (let j = i + 1; j < words.length; j++) {
      if (isPalidrome(words[i], words[j])) {
        res.push([i, j]);
      }

      if (isPalidrome(words[j], words[i])) {
        res.push([j, i]);
      }
    }
  }

  return res;

  function isPalidrome(a = '', b = '') {
    let i = 0;
    let j = b.length - 1;

    while (i < a.length && j >= 0) {
      if (a[i] != b[j]) {
        return false;
      }
      i++;
      j--;
    }

    let k = a.length - 1;

    while (i < a.length) {
      if (a[i] != a[k]) return false;
      i++;
      k--;
    }

    k = 0;

    while (k < j) {
      if (b[k] != b[j]) return false;
      j--;
      k++;
    }

    return true;
  }
};
